### ISSUE TITLE
Feature: A Package Panel (Tab) for the QuickStart Wizard

### ISSUE THREAD

**Issue by octaeder** [texstudio-org/texstudio#2831]
For all: I want to let you know about an almost finished work. If you like tell me which packages should I add to the packages list.

### Introduction

The following image shows the QuickStart Wizard as you can find it in txs. I added some marks for later reference:

![grafik](https://user-images.githubusercontent.com/102688820/211026502-f6d75210-e978-46f6-8fc5-20ca205c3090.png)

Even so the tab claims "Class Options" there are other things included (marks 1 to 4). The first two marks show commonly used options (if you don't use LuaLaTeX) for packages but which integrate well into the tabs geometry. This is also true for the babel option/package. But the following line with three options (mark 4) are packages not options. And the first of them (AMS) in fact represents two packages, but which packages is not clear at first (amsmath and amssymb). This line of option doesn't fit well into the geometry and leads to a scrollbar (mark 5) because the "Other Options" widget has not enough space. As a side note: The geometry package has it's own tab.

### A new Packages Tab

So the idea was born to have a separate "packages" tab widget having a table of (about 30) useful packages with a small description that can help one figure out if this package could be of interest. The four packages (mark 4) are moved there. Some packages have options that are assumed to be useful and are also shown (and automatically used when creating the document). As before txs stores the package names selected for later use.

The following image shows the new QuickStart dialog right after starting it with the new Packages tab:

![grafik](https://user-images.githubusercontent.com/102688820/211030089-fa8a0e7f-d720-4d8c-8812-4e7e0d455a8f.png)

Activation of that tab also activates a tool tip for this tab widget:

![grafik](https://user-images.githubusercontent.com/102688820/211031511-c5dea337-2ca4-4fdc-887b-56172a754b48.png)

### The List of Packages

The following images show the complete list of packages:

![grafik](https://user-images.githubusercontent.com/102688820/211032511-a4c71006-22ea-4906-99e0-3f903a434848.png)
![grafik](https://user-images.githubusercontent.com/102688820/211034097-f7b29492-3f08-41dc-9e2a-0629aa75220d.png)
![grafik](https://user-images.githubusercontent.com/102688820/211034836-c588049d-b556-49dc-a755-520a68abf18a.png)

As mentioned in the tool tip you can select packages with check boxes or by selecting lines:

![grafik](https://user-images.githubusercontent.com/102688820/211035244-b3bf8839-fb94-414a-9d60-f337cddf21a7.png)

Thus you may even select all packages within a second. When you need line numbers for better orientation you may activate them with the check box at the bottom:

![grafik](https://user-images.githubusercontent.com/102688820/211035738-41290445-cb54-433c-9503-744a249eab13.png)

And here's the final document:

![grafik](https://user-images.githubusercontent.com/102688820/211035928-ea35579f-c15b-4a6d-8b4f-695ab79e14e5.png)

**Comment by sunderme** [comment]
I see the point, but I also don't like the already long list of packages in the extra tab.
The point of quick start is **quick** start and scrolling down a long list (of unknown) packages opposes that idea.
In my experience, users set-up their paper/article etc. style once and should use personal templates from that point on.
After all, it is easy to add packages in text, there is no big benefit doing so with the gui. 
So, in my opinion, the dialog should give a first usable template with some flexibility (languages) which is as easy (non-threatening to a beginner) as possible.
The number of packages and the option column does not fulfill this.

**Comment by sunderme** [comment]
I would actually take out utf8 input enc (and set it always)
Is T1 fontenc really needed (anymore)
And the one additional package which is very likely is "graphicx"
AMSpackage maybe, the makeindex is probably already rarely used.

**Comment by mbertucci47** [comment]
utf8 has been the default input encoding since 2018 so `\usepackage[utf8]{inputenc}` doesn't do anything with a somewhat up-to-date LaTeX (which I know can't always be assumed). This is explained in [ltnews28](https://www.latex-project.org/news/latex2e-news/ltnews28.pdf).

**Comment by sunderme** [comment]
ubuntu 18.04 support is retired in April, so after that latex can be assumed to be newer.

**Comment by octaeder** [comment]
Yes, this package/options list is totally over engineered. Be aware that this list is not just an alphabetical list, it groups related stuff and sorts expert stuff to the end. So for a first time quick start she should find her packages at the top. Even so this list is surely not perfect as the layout of the Class Options tab isn't either. So I would expect discussing what could be done better. Throw away the options column, cut down or reorder the list, etc. This would still suffice reaching the point I'm targeting to: Give the user some orientation where she could go next in the jungle of packages. Hints are given by the descriptions.

If one doesn't care about packages, perfect, don't go the packages tab. But currently the user sees those AMS, makeindex and graphics options on the first tab and starts thinking what they are about.

Following your approach I would remove the Geometry tab for several reasons and remove all add-to buttons. BTW these buttons don't follow the icon theme.

Whatever your decision, I can only continue to work on this wizard with more specific requirements. So let me know.

**Comment by sunderme** [comment]
okay requirements (suggestions):
1) by default all offered packages should fit in the tab without scrolling
2) the offered packages should be beginner oriented (so rather basic)
3) the presentation should be simple (i.e. no options column)
4) an extended/advanced button/mode or similar may be available

I would keep the geometry tab (because there may be users and the graphical feed-back is actually nice).

**Comment by octaeder** [comment]
Is the layout ok (s. image)? Notes: I tried many things to have the tableWidget exactly the size needed without any scrollbar. The only way I found to do so is to sum up the heights of the header and all rows and fix the widget's height to this sum. Thus the window height can only shrink as long as there is enough space for the full table. A related problem is that the sum may not suffice as height depending on style settings (or OS?). So a scrollbar may appear because the size is one or two pixels to small.

![grafik](https://user-images.githubusercontent.com/102688820/211221313-8588d194-8bdf-40b8-b17e-3371e45f9cac.png)

Since you said that the graphicx package is the most "important" I moved it to the top. Please be aware of the reference to the Insert Grahic Wizard in the description for the graphicx package.
I added a copy package(s) button which copies all \usepackage commands for the selected packages to the clipboard. Just in case you want to past it in your own document.

What about the utf8 and T1 encoding? Is there any decision? I understood use always utf8 encoding, so I could remove the combobox.

PS: The first post didn't mention that I added key accelerators for the tabs (ALT+C/P/G).

**Comment by sunderme** [comment]
I agree, beginner type packages typically look different. (there seems to be  a bias for the author's setup ;-) )

I probably would sort by letter, not by arbitrary "importance" because someone has to pick the ones he wants later (and should not read every line)

The point of the table height is to have few enough item by default and a scrollbar once they become to many (which is fine then)

"I added a copy package(s) button which copies all \usepackage commands for the selected packages to the clipboard. Just in case you want to past it in your own document." 
Not sure what should be the use-case for this ?

not inputenc, but T1 fontenc should stay.

**Comment by sunderme** [comment]
@tmelorc , all, any suggestions what "beginner" packages are/should be ?

**Comment by tmelorc** [comment]
> @tmelorc , all, any suggestions what "beginner" packages are/should be ?

Well, usually for my students I recommend `amsmath`, `amthm`, `amsfonts`, `geometry`, `graphics`, `babel`, `enumitem` and (maybe) `xcolor`. 

Since they are new users, too complicated packages could lead them to go to Google and copy/paste a lot of complexes codes.

**Comment by dbitouze** [comment]
I would recommend:

- `mathtools` instead of  `amsmath` (the former loading the latter),
- `graphicx` instead of  `graphics`,
- `hyperref` which is often wanted (even by users who do not suspect such features),
- to not load `enumitem` since:
  - it is hasn't any effect if it is just loaded,
  - it encourages users to “try to emulate what they know no matter what.” (Raphael on TeX.SE).

**Comment by dbitouze** [comment]
IMO, QuickStart Wizard is much less useful than templates and I insist again on a feature that I find interesting: #775

**Comment by octaeder** [comment]
Do LaTeX beginners use packages? Thanks for the discussion.

I'll
- remove extra button
- stay with graphicx: The Insert Graphic Wizard will complain when no \usepackage{graphicx} is in the preamble. See description in the list.
- bring back hyperref (but without options)
- skip hhline, longtable, xfrac, rotating, wrapfig, tikz, and siunitx
- stay with mathtools, enumitem, nameref, and xcolor
- keep ntheorem
- font handling for beginners (amsfonts)? I would like to here from others.
- geometry and babel are not part of the discussion (s. tabs Class Options and Geometry)

Comment on ordering the list: _If the amsthm package is used with a non-AMS document class and with the amsmath package, amsthm must be loaded after amsmath, not before._ So I would have to place amsthm out of order. Since amsthm enhances command \newtheorem I doubt if defining new environments is for beginners.

A word on enumitem, geometry, and nameref: Following typesetting rules maybe one should not use geometry. But enumitem can simplify some tasks. Referencing titles of chapters and sections is common with with office writers. I think the user should have the possibility to use it.

![grafik](https://user-images.githubusercontent.com/102688820/211389333-ceef4bce-7298-43b2-be7c-72f9180438d8.png)

So like this?

**Comment by sunderme** [comment]
I would not really encourage to use enumitem. Typically that kind of thing should be defined in the class.

**Comment by dbitouze** [comment]
>     * stay with mathtools, enumitem, nameref, and xcolor

See below.
 
>     * keep ntheorem

It isn't maintained anymore and is often not recommended. `amsthm` looks preferable but, IMO, `thmtools` should be encouraged.

>     * font handling for beginners (amsfonts)? I would like to here from others.

`amssymb`  looks more useful (and load `amsfonts`).

> But enumitem can simplify some tasks.

But encourages bad typographic practices for those who are not aware.

> Referencing titles of chapters and sections is common with with office writers. I think the user should have the possibility to use it.

`nameref` is loaded by `hyperref`.

**Comment by octaeder** [comment]
seems that thmtools needs amsthm (see all examples of chapter 1 in the thmtools documentation).
maybe you want to use nameref without hyperref, so I made a note

So current state is:

![grafik](https://user-images.githubusercontent.com/102688820/211426782-4acc3522-0d13-43d2-891d-8f33207994a5.png)

**Comment by octaeder** [comment]
Change to first tab (Class Options) as suggested previously: I removed line with Input encoding (image mark 1), but didn't touch line with Font encoding. This results in:

![grafik](https://user-images.githubusercontent.com/102688820/211618042-f0235715-4486-4e71-a80c-b7c5e522d180.png)

The code now always insert `` \usepackage[utf8]{inputenc}``  to the document.

I also removed code inserting `` usepackage{ucs}`` which was used in case of "utf8x".

Since all words in labels here start with a capital I will switch to Font Encoding.
